csi/server/plugin/opensds/controller.go

```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
 					},
 				},
 			},
+			&csi.ControllerServiceCapability{
+				Type: &csi.ControllerServiceCapability_Rpc{
+					Rpc: &csi.ControllerServiceCapability_RPC{
+						Type: csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT,
+					},
+				},
+			},
 		},
 	}, nil
 }
 
-// CreateSnapshot
+// CreateSnapshot implementation
 func (p *Plugin) CreateSnapshot(
 	ctx context.Context,
 	req *csi.CreateSnapshotRequest) (
 	*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+
+	defer glog.Info("end to CreateSnapshot")
+	glog.Infof("start to CreateSnapshot, Name: %v, SourceVolumeId: %v, CreateSnapshotSecrets: %v, parameters: %v!",
+		req.Name, req.SourceVolumeId, req.CreateSnapshotSecrets, req.Parameters)
+
+	if 0 == len(req.Name) {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot Name cannot be empty")
+	}
+
+	if 0 == len(req.SourceVolumeId) {
+		return nil, status.Error(codes.InvalidArgument, "Source Volume ID cannot be empty")
+	}
+
+	client := sdscontroller.GetClient("", "")
+	snapReq := &model.VolumeSnapshotSpec{
+		Name:     req.Name,
+		VolumeId: req.SourceVolumeId,
+	}
+
+	snapshot, err := client.CreateVolumeSnapshot(snapReq)
+	if nil != err {
+		return nil, err
+	}
+
+	createdAt, err := time.Parse(constants.TimeFormat, snapshot.CreatedAt)
+	if nil != err {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      snapshot.Size,
```

Comment:
Size seems to be wrong from your video.  You created a "1Gi" volume specified in pvc_sc.yaml, but "kubectl describe volumesnapshot" showed "Restore Size: 1".  That means 1 byte?   It should be "Restore Size:  1Gi".